src/compute-client/src/controller.rs

```diff
@@ -228,7 +228,7 @@ impl<T: Timestamp> ComputeController<T> {
     ) -> Result<&CollectionState<T>, CollectionLookupError> {
         self.instances
             .values()
-            .flat_map(|i| i.collection(collection_id).ok())
+            .flat_map(|i| i.live_collection(collection_id).ok())
```

Comment:
Is it ok that `find_collection` will now sometimes not find a collection that is still in the Controller state? I can see `install_watch_set` calling this, which is called from `peek_stage_finish`on the peek's dependencies. Do we have a guarantee that a peek won't depend on an object that is `dropped` but is still in the Controller state? I'm not so sure, especially considering that peek sequencing happens not fully on the main thread nowadays. I guess the question is whether we prevent the dropping of such objects for which there is a peek sequencing in progress, or is the peek sequencing expected to gracefully handle the case that the dependencies suddenly disappear during peek sequencing.

Reply:
The commit introduces the new assumption that once you have dropped a collection with the controller API you have no reason to expect to find it anymore. I think that's not a crazy assumption to have, but you might be right that it doesn't work well with how we handle peeks.

[We do have code](https://github.com/teskje/materialize/blob/55017d7d67078a35cc5942e22961dc377f65f63a/src/adapter/src/coord/sequencer/inner/peek.rs#L205-L212) that ensures that in `peek_stage_finish` all dependencies are always still there. I'm not sure how much we should rely on it though. In theory, peeks still work after you dropped their dependencies, as long as you required the appropriate read holds first.

I made the change to fix some test where adapter was listing the reverse dependencies of a collection, found one it had already dropped from the catalog and then got confused by not knowing what this dependency is. Should be easy to change that in adapter instead, so we don't have to change our behavior that dramatically.